### DISCUSSION TITLE
удаление квестовых предметов у торговцев

### DISCUSSION THREAD

**Discussion by Kor-Angar** [dosinabox/g2nr_unofficial_update#465] [General]
а можно в скриптах где удаляются шмотки у торговцев добавить проверки:

- [x] выполнен или квест Секстант(_MIS_RamirezSextant_) и если выполнен удалить секстант(_ItMi_Sextant_)
- [x] золотое блюдо Бромора 
- [ ] жир кротокрыса проверка FoundVinosKellerei
- [x] голова Бладвина
- [x] шмотки Грега
- [x] амулет Грега
- [ ] кольцо Константино
- [ ] цветные таблички  проверка на сдачу квеста Сатураса или смерть Равена
- [x] кровавые чаши, квест на кровавые чаши
- [x] золотое блюдо Марии
- [ ] драгоценная статуэтка Инноса
- [ ] кольцо гильдии, улика против Фернандо
- [ ] Божественная сила звезд

**Comment by Kor-Angar** [comment]
лук Боспера квестовый?
арбалет паладина в 4 главе? 
арбалет Драгомира? 
меч Онд в 4 главе?

**Comment by dosinabox** [comment]
> лук Боспера квестовый?
> меч Онд в 4 главе?

да

> арбалет паладина в 4 главе?
> арбалет Драгомира?

нет

**Comment by dosinabox** [comment]
> * шмотки Грега
> * амулет Грега

не обязательно, потому что теперь их можно вернуть в Яркендаре, получив меньше опыта

**Comment by Kor-Angar** [comment]
> не обязательно, потому что теперь их можно вернуть в Яркендаре, получив меньше опыта

когда в Ярике шмотки сдаются квест закрывается?

**Comment by dosinabox** [comment]
> когда в Ярике шмотки сдаются квест закрывается?

При переходе в Яркендар квест пропадает из дневника (отменяется).

**Comment by Kor-Angar** [comment]
> При переходе в Яркендар квест пропадает из дневника (отменяется).

тогда можно сделать проверку на диалог c Грегом в Ярике. так то Грег в Хориносе бессмертный у него его шмотки отобрать не получится :-)

**Comment by Kor-Angar** [comment]
кольца Тенгрона?

**Comment by Kor-Angar** [comment]
дневник корнелиуса
письмо диего

**Comment by dosinabox** [comment]
> дневник корнелиуса

Его забирает Хаген

> письмо диего

Его забирает Гербрандт

**Comment by Kor-Angar** [comment]
![image](https://user-images.githubusercontent.com/49514764/95631335-3375f600-0a8c-11eb-9582-1583bf77ebfb.png)
"рецепты, книга, портреты ГГ, кольцо от портала, каменные таблички, 6 кровавых чаш, голова друга"

**Comment by dosinabox** [comment]
кем убит?

**Reply by Kor-Angar** [reply]
героем избит

**Comment by Kor-Angar** [comment]
а почему чаши кровавые не удалились?

**Comment by Kor-Angar** [comment]
давай все таблички каменные отдавать Сатурасу в диалоге про убийство Ворона.
давай с портретов ГГ снимет метку квестовых. они же никакой роли не играют вообще

**Comment by dosinabox** [comment]
> а почему чаши кровавые не удалились?

Из-за квестового флага

> давай с портретов ГГ снимет метку квестовых. они же никакой роли не играют вообще

Так у них и цена нулевая

**Comment by Kor-Angar** [comment]
я думал ты в шапке галочки поставила там где исправила :-)

**Comment by dosinabox** [comment]
да, и тогда странно, потому что такая проверка уже есть:

B_ClearJunkTradeInv.d:
```cpp
if(MIS_CassiaKelche == LOG_SUCCESS)
{
	Npc_RemoveInvItems(slf,ItMi_BloodCup_MIS,Npc_HasItems(slf,ItMi_BloodCup_MIS));
};
```

Хотя не странно, потому что квест может быть провелен. Добавлю и такую проверку.

**Comment by Kor-Angar** [comment]
так я квест выполнил про кровавые чаши

**Comment by Kor-Angar** [comment]
вполнил квест. сдал чашки Кассии. потом по квесту Андре ликвидировал гильдию.

**Comment by Kor-Angar** [comment]
можно просто делать проверку на смерть Кассии для кольца Констанино и чашам кровавым.

**Comment by Kor-Angar** [comment]
или такое будет работать только в Хориносе?

**Comment by Kor-Angar** [comment]
> Так у них и цена нулевая

так это же хлам, если игрок его продал кому то, то наверное ему это не надо :-)

**Comment by dosinabox** [comment]
> вполнил квест. сдал чашки Кассии. потом по квесту Андре ликвидировал гильдию.

Ликвидация гильдии устанавливает все квесты воров в LOG_FAILED (но не меняет записи в дневнике). Вот чаши и не удалились.

> или такое будет работать только в Хориносе?

да

**Comment by Kor-Angar** [comment]
> Ликвидация гильдии устанавливает все квесты воров в LOG_FAILED (но не меняет записи в дневнике). Вот чаши и не удалились.

исправлять будешь? :-)

**Comment by dosinabox** [comment]
> исправлять будешь? :-)

уже

**Comment by Kor-Angar** [comment]
квестовый лут у Халвора
![20210118204511_1](https://user-images.githubusercontent.com/49514764/105000797-dd77fd80-5a3f-11eb-88cf-af4d26bf5d72.jpg)

**Comment by Kor-Angar** [comment]
можно просто снимать флаг миссии у квестовых предметах когда герой их отдаёт квестодателю в функции  B_GiveInvItems

```D
if(item.flags & ITEM_MISSION)
	{
		item.flags = item.flags & ~ITEM_MISSION;
	}
```

**Comment by Kor-Angar** [comment]
предлагаю попробовать потестить

**Comment by Kor-Angar** [comment]
можно сделать опционально например. я в этом прохождение, что сейчас полноценно протестирую.

```D
if(Npc_IsPlayer(giver))
	{
		Npc_GetInvItem(giver,itemInstance);
		itemname = item.description;
		if(itemInstance == ItMi_Gold)
		{
			concatText = ConcatStrings(IntToString(amount),PRINT_GoldGiven);
			AI_PrintScreen(concatText,-1,YPOS_GoldGiven,FONT_ScreenSmall,2);
		}
		else if(amount == 1)
		{
			concatText = ConcatStrings(itemname,PRINT_ItemGiven);
			AI_PrintScreen(concatText,-1,YPOS_ItemGiven,FONT_ScreenSmall,2);
		}
		else
		{
			concatText = ConcatStrings(IntToString(amount),PRINT_ItemsGiven);
			concatText = ConcatStrings(concatText," (");
			concatText = ConcatStrings(concatText,itemname);
			concatText = ConcatStrings(concatText,")");
			AI_PrintScreen(concatText,-1,YPOS_ItemGiven,FONT_ScreenSmall,2);
		};
		Npc_RemoveInvItems(giver,itemInstance,amount);
		CreateInvItems(taker,itemInstance,amount);
		if(item.flags & ITEM_MISSION)
		{
			item.flags = item.flags & ~ITEM_MISSION;
		};
	}
```

**Comment by dosinabox** [comment]
Возможно, нужно itemInstance.flags вместо item.flags.
Возможно, изменения флагов не сохранятся в сохраненку.

**Reply by Kor-Angar** [reply]
Я исходил из того что у тебя в коде выше идет дискрипшион.айтем Про сохранения, заодно и проверим ;-)

**Comment by dosinabox** [comment]
item - конкретный единичный экземпляр класса itemInstance. Брать у него свойство - нормально. Назначать флаги - не факт.
Кстати, нужно создавать предмет у получателя только после назначения флагов. Даже если изменение флагов класса не затронет все существующие предметы или не пойдет в сохраненку, то у получателя останется вновь созданный экземпляр с измененными флагами. Скорее всего.
И удалять предмет тоже следует до изменения флагов, иначе проверка свойств удаленного объекта может не сработать.

```cpp
if(itemInstance.flags & ITEM_MISSION)
{
	itemInstance.flags = itemInstance.flags & ~ITEM_MISSION;
};
Npc_RemoveInvItems(giver,itemInstance,amount);
CreateInvItems(taker,itemInstance,amount);
```

**Reply by Kor-Angar** [reply]
Надо добавить в тестовую сборку 😀 потестить 😀Заодно и узнаем на примере табличек которые Сатурасу сдаются, как они себя вести будут, после того как  Сатурас их назад  герою вернёт.

**Comment by dosinabox** [comment]
Окей, сейчас упакую и выложу.
```cpp
if(item.flags & ITEM_MISSION)
{
	item.flags = item.flags & ~ITEM_MISSION;
	AI_PrintScreen(ConcatStrings(itemname, " - флаг снят"),-1,YPOS_GoldTaken,FONT_ScreenSmall,2);
};
```
![Screenshot_1](https://user-images.githubusercontent.com/39790055/117568693-6e93e280-b0ca-11eb-9e9e-567fb5225eb8.jpg)

**Comment by Kor-Angar** [comment]
Я просто не уверен что такое решение решит вопрос в целом. Слишком оно простое и очевидное 😀 поэтому хочу пройти игру что бы сказать, что были вот такие негативные моменты и оно не подойдёт

**Comment by Kor-Angar** [comment]
еще бы с карт снять "квестовасть". а продавцу Ибрагиму карты просто добавлять в продажу если у него их нет.

**Comment by Kor-Angar** [comment]
вчера следил за этой историей. работает корректно. стараюсь продавать квестовые предметы паре продавцов, которые по сюжету получать люлей :-)

**Comment by Kor-Angar** [comment]
```D
if(item.flags & ITEM_MISSION)
{
	item.flags = item.flags & ~ITEM_MISSION;
	AI_PrintScreen(ConcatStrings(itemname, " - флаг снят"),-1,YPOS_GoldTaken,FONT_ScreenSmall,2);
};
```

это добавлять в основной код не стала? :-)

**Comment by dosinabox** [comment]
> это добавлять в основной код не стала? :-)

нет, убрала к какому-то из релизов

**Comment by Kor-Angar** [comment]
очень хорошая штука. найди возможность поставить хотя бы опционально. решение получилось крайне элегантным и эффективным.

**Comment by dosinabox** [comment]
> очень хорошая штука. найди возможность поставить хотя бы опционально. решение получилось крайне элегантным и эффективным.

Насколько эффективным? Точно стоит добавлять?

**Comment by Kor-Angar** [comment]
Эффеткивно на все 100%. Точно стоит.

**Comment by Kor-Angar** [comment]
это штука эффективна на 300%. код из 1 строчки решает целый комплекс задач:

* пресекается возможность спекуляций: по несколько раз продавать одни и те же вещи, таким образом "зарабатывая" золото
* тем кто против спекуляций тоже будет проще, так как продажа квестовых предметов не будет отличаться от продажи обычных предметов
* уже в 4ой главе квестовых предметов скапливается достаточно много и они объективно заполняют инвентарь, при этом при продаже их мы возвращаемся к проблеме что они остаются у торговца. таким образом эти предметы становятся какими то "особенными", хотя таковыми не являются
* идеологический вопрос. флаг для миссия был создан, на мой взгляд, в первую очередь что бы предметом не мог пропасть из мира, что бы игрок имел возможно закрыть квест, даже в том случае, если случайно продал данный предмет. после завершения миссии предмет теряет свой особый статус и он становится обычным. собственно это строчка именно это и делает: лишает предмет такого статуса.

конечно можно пойти еще дальше и снимать флаг в B_CloseTopic.d. но это уже будет избыточно и слишком грамоздко.

со снятием флага я прошёл игру два или три разу. внимательно отслеживал как это происходит. 

возможно в игре где то есть такой момент когда НПС берёт у героя квестовый предмет смотрит его, а потом возвращает обратно, но это в первую очередь касается документов. можно добавить дополнительную проверку на документы и с них не снимать флаги для миссии. но и это будет избыточно на мой взгляд :-)

**Comment by dosinabox** [comment]
> Эффеткивно на все 100%. Точно стоит.

Добавлено в сборку для мастерской Стима.

**Reply by Kor-Angar** [reply]
https://github.com/dosinabox/g2nr_unofficial_update/blob/b3d384db96087b4cb577714874115bb91c170c68/PrjGOTHIC/Story/B_Story/B_GiveInvItems.d#L21

**Comment by Kor-Angar** [comment]
@dosinabox я выложу в стим ввиде отдельного мини-мода? ты не против?

**Reply by dosinabox** [reply]
не против

**Reply by Kor-Angar** [reply]
млин B_GiveInvItems это func **int** и не получиться инъекцию сделать в конце  старой функции :-(

**Comment by Kor-Angar** [comment]
как то так. вроде работает

```D
var string DIA_CurrentName;

func int B_GiveInvItems (var C_NPC giver, var C_NPC taker, var int itemInstance, var int amount)
{
    // EXIT IF...

    if (Npc_IsPlayer(giver))
    {
        if (amount > Npc_HasItems(giver,itemInstance))
        {
            return FALSE;
        };
    };
    
    if (amount == 0)
    {
        return TRUE;
    };
    
    //FUNC
  
    Npc_GetInvItem(giver,itemInstance);

    var string msg;   
    msg = Str_Format("[%s] --> ([%s] --> [%s])  B_GiveInvItems[%s]",DIA_CurrentName, giver.name, taker.name ,item.description);
    Hlp_PrintConsole(msg);
             
    if  Npc_IsPlayer(giver)
    {
        //Hlp_PrintConsole("Npc_IsPlayer(giver)");
        if (item.flags & ITEM_MISSION)
        {
            item.flags = item.flags & ~ITEM_MISSION;
            msg = Str_Format("[%s] -->  [%s] --> (item.flags & ~ITEM_MISSION)",DIA_CurrentName, item.description);
            Hlp_PrintConsole(msg);
        };
    };  
    
    B_GiveInvItems_old(giver, taker, itemInstance, amount);  
};
```

**Reply by Kor-Angar** [reply]
или нет :-)  ![20220728212742_1](https://user-images.githubusercontent.com/49514764/181612706-9c962e70-2045-49af-b7a7-afd79e998de4.jpg)

**Reply by Kor-Angar** [reply]
ну а с другой стороны я же просто могу полностью заменить оригинальную функцию вставив кусочек про снятие флага :-)
[B_GiveInvItems.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/9214505/B_GiveInvItems.zip)

**Reply by Kor-Angar** [reply]
https://steamcommunity.com/sharedfiles/filedetails/?id=2841743939
